daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala

```diff
@@ -67,24 +95,155 @@ class EvaluationOrderTest extends AnyWordSpec with Matchers {
+                let c: M:T = M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 0 }
+                in sbind x : ContractId M:T <- create @M:T c
+                  in create @M:T c
+           """,
+          Array(SParty(alice), SParty(bob)),
+          alice,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[DuplicateContractKey]
+        }
+        msgs shouldBe allTraces ++ allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with empty contract key maintainers
+      "empty contract key maintainers" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:keyNoMaintainers, nested = M:buildNested 0 }
+           """,
+          Array(SParty(alice), SParty(bob)),
+          alice,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[CreateEmptyContractKeyMaintainers]
+
+        }
+        msgs shouldBe allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with authorization failure
+      "authorization failure" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 0 }
+           """,
+          Array(SParty(alice), SParty(bob)),
+          bob,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[FailedAuthorization]
+
+        }
+        msgs shouldBe allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with contract id in contract key
+      "contract id in contract key" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) (cid : ContractId Unit) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:toKeyWithCid sig cid, nested = M:buildNested 0 }
+           """,
+          Array(
+            SParty(alice),
+            SParty(bob),
+            SContractId(ContractId.V1.assertFromString("00" * 32 + "0000")),
+          ),
+          alice,
+        )
+        inside(res) { case SResultError(SErrorDamlException(err)) =>
+          err shouldBe a[ContractIdInContractKey]
+
+        }
+        msgs shouldBe allTraces
+      }
+      // TEST_EVIDENCE: Semantics: Evaluation order of create with create argument exceeding max nesting
+      "create argument exceeds max nesting" in {
+        val (res, msgs) = evalUpdateApp(
+          pkgs,
+          e"""\(sig : Party) (obs : Party) -> create @M:T
+              M:T { signatory = sig, observer = obs, precondition = True, key = M:toKey sig, nested = M:buildNested 100 }
```

Comment:
Do we / should we test the precise value of this nesting limit? 49 ok; 50 blows.

Reply:
We have those tests in other places. Here I’m only interested in ordering not in testing that the checks themselves are fine.

Reply:
:+1: